wire_protocol/src/connection.rs

```diff
@@ -227,7 +237,13 @@ impl Connection for TlsConnection {
     }
 
     fn is_valid(&self) -> bool {
-        self.stream.as_ref().expect("get connection").is_valid()
+        self.is_valid
+            && self.stream.as_ref().is_some()
+            && self.stream.as_ref().expect("get connection").is_valid()
+    }
+
+    fn set_validity(&mut self, is_valid: bool) {
```

Comment:
A comment might be in order, because this leads to the somewhat surprising case that calling `set_validity(true)` followed by `is_valid` does not always return `true`.